crates/rattler_installs_packages/src/index/package_database.rs

```diff
@@ -111,6 +127,390 @@ impl PackageDb {
+                    .await?;
+                (wheel_metadata.0, wheel_metadata.1, name)
+            } else {
+                let (wheel_metadata, name) = self
+                    .get_stree_from_file_path(
+                        &normalized_package_name,
+                        url.clone(),
+                        None,
+                        wheel_builder,
+                    )
+                    .await?;
+                (wheel_metadata.0, wheel_metadata.1, name)
+            };
+
+        let artifact_hash = {
+            ArtifactHashes {
+                sha256: Some(rattler_digest::compute_bytes_digest::<Sha256>(
+                    metadata_bytes.clone(),
+                )),
+            }
+        };
+
+        let artifact_info = ArtifactInfo {
+            filename: artifact_name,
+            url: url.clone(),
+            hashes: Some(artifact_hash),
+            requires_python: metadata.requires_python,
+            dist_info_metadata: DistInfoMetadata::default(),
+            yanked: Yanked::default(),
+        };
+
+        let mut result: IndexMap<PypiVersion, Vec<ArtifactInfo>> = Default::default();
+
+        result
+            .entry(PypiVersion::Url(url))
+            .or_default()
+            .push(artifact_info.clone());
+
+        self.put_metadata_in_cache(&artifact_info, &metadata_bytes)
+            .await?;
+
+        Ok(self
+            .artifacts
+            .insert(normalized_package_name.clone(), Box::new(result)))
+    }
+
+    /// Return an sdist from file path
+    pub async fn get_sdist_from_http<'a, 'i>(
+        &self,
+        normalized_package_name: &NormalizedPackageName,
+        url: Url,
+        bytes: Box<dyn ReadAndSeek + Send>,
+        wheel_builder: &WheelBuilder<'a, 'i>,
+    ) -> miette::Result<((Vec<u8>, WheelCoreMetadata), ArtifactName)> {
+        // it's probably an sdist
+        let distribution = PackageName::from(normalized_package_name.clone());
+        let version = Version::from_str("0.0.0").expect("0.0.0 version should always be parseable");
+        let format = SDistFormat::get_extension(url.path()).into_diagnostic()?;
+
+        let dummy_sdist_file_name = SDistFilename {
+            distribution,
+            version,
+            format,
+        };
+
+        let dummy_sdist = SDist::new(dummy_sdist_file_name, Box::new(bytes))?;
+
+        let wheel_metadata = wheel_builder
+            .get_sdist_metadata(&dummy_sdist)
+            .await
+            .into_diagnostic()?;
+
+        // construct a real sdist filename
+        let sdist_filename = SDistFilename {
+            distribution: wheel_metadata.1.name.clone(),
+            version: wheel_metadata.1.version.clone(),
+            format,
+        };
+        let filename = ArtifactName::SDist(sdist_filename.clone());
+
+        Ok((wheel_metadata, filename))
+    }
+
+    /// Get artifact by file URL
+    pub async fn get_http_artifact<'a, 'i, P: Into<NormalizedPackageName>>(
```

Comment:
Name confuses me a bit, if it is for direct_url references I would opt for something like `get_direct_url_artifact` or `get_explicit_url_artifact` or something.